ships/0031-shipwright-trigger.md

```diff
@@ -0,0 +1,296 @@
+Shipwright Trigger
+------------------
+
+# Summary
+
+Shipwright's project goal is to automate and greatly simplify the container image building process, and currently, users can only initiate the build process by issuing a new `BuildRun` object.
+
+This enhancement proposal describes how Shipwright will react upon WebHook events coming from Git service providers like GitHub, how it will be incorporated into Tekton Pipelines via Custom Task, and how it will react upon Tekton Pipelines resources.
+
+# Motivation
+
+An essential part of the developer experience is to be able to commit and push code changes, while rest assured that will actionate automation procedures to produce a new container image.
+
+As well as, being able to combine the image creation process with Continuous Integration pipelines, and watch over Tekton Pipelines which may be producing dependencies and other artifacts used on container images built by Shipwright.
+
+Thus, the primary motivation for this enhancement proposal is to improve developer experience using Shipwright, by providing a good level of integration with Git service providers and Tekton Pipelines.
+
+## Goals
+
+- Improve developer experience
+- Allow integration of Git service providers via WebHook
+- Incorporate into Tekton Pipelines via Custom Tasks
+- integration with Tekton Pipelines, being able to react upon Pipeline executions
+
+## Non-Goals
+
+- General Git service providers WebHook support (only GitHub supported initially)
+- Modify `.spec.output` directives
+
+# Proposal
+
+Introduce a new project name Shipwright Trigger (`shipwright-io/trigger`) to based on the rules defined on the Build resources, listen to WebHook events coming from Git service providers, incorporate Shipwright into Tekton via `CustomTask` (Tekton Run resource), and watch over Tekton Pipeline resources in order to trigger `BuildRun` resources accordingly.
+
+## API Changes
+
+We will modify the `Build` CRD to allow users to define how to trigger it. The API will look like the following:
+
+```yaml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+spec:
+  source:
+    url: https://github.com/otaviof/nodejs-ex.git
+  trigger:
```

Comment:
What about making this a resource on its own? IIUC the way things are designed here it seems like we would be introducing fields to the `Build` struct that would be meaningful only to this new `trigger` project (or operator/controller if I may).

This is just food for thought: what about having a new type called `BuildTrigger` that refers to a `Build` by its name and then have this new type living only in the new `trigger` repo ? I am pretty new to the project so please educate me on why this is not a good idea.

Reply:
> What about making this a resource on its own? IIUC the way things are designed here it seems like we would be introducing fields to the `Build` struct that would be meaningful only to this new `trigger` project (or operator/controller if I may).

There's an intimate relationship between the `Build` resource and the Git repository itself. For instance, for GitHub WebHook you can employ a shared secret between the repository and the Shipwright Build, in order to validate the caller, increase security.

So, the attribute `.spec.trigger.secretRef` can be unique per `Build`, and in case it's shared between more resources, it would be shared via the `Secret` resource (name).

The `.spec.trigger.when[]` could be common for more `Build` resources, indeed. However, I still look at it as means to trigger that specific `Build`, so once again an "intimate relationship".